crates/fj-core/src/validate/shell.rs

```diff
@@ -324,24 +329,56 @@ impl ShellValidationError {
             errors.push(Self::NotWatertight.into());
         }
 
-        let mut half_edge_to_faces: HashMap<ObjectId, usize> = HashMap::new();
+        let mut global_edge_to_faces: HashMap<ObjectId, usize> = HashMap::new();
 
         for face in shell.faces() {
             for cycle in face.region().all_cycles() {
                 for half_edge in cycle.half_edges() {
                     let id = half_edge.global_form().id();
-                    let entry = half_edge_to_faces.entry(id);
+                    let entry = global_edge_to_faces.entry(id);
                     *entry.or_insert(0) += 1;
                 }
             }
         }
 
         // Each global edge should have exactly two half edges that are part of
         // the shell
-        if half_edge_to_faces.iter().any(|(_, c)| *c != 2) {
+        if global_edge_to_faces.iter().any(|(_, c)| *c != 2) {
             errors.push(Self::NotWatertight.into())
         }
     }
+
+    fn validate_same_orientation(
+        shell: &Shell,
+        errors: &mut Vec<ValidationError>,
+    ) {
+        let mut global_to_half: HashMap<ObjectId, Vec<_>> = HashMap::new();
```

Comment:
For some additional type safety, it should be possible to replace `ObjectdId` with `HandleWrapper<GlobalEdge>` here. This was not possible before, due to a horribly shameful workaround (:grin:), but that workaround was removed in https://github.com/hannobraun/fornjot/pull/1951.

This is just a note, not a change request. I don't think it would make a difference here.